In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import * 
import matplotlib.pyplot as plt

from sklearn.metrics import mean_absolute_error, mean_squared_error, median_absolute_error, r2_score
from sklearn.preprocessing import MinMaxScaler
from sklearn.neighbors import KNeighborsRegressor
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler

In [2]:
data = pd.read_csv('../datasets/heart.csv')

In [3]:
dataCopy = data.copy()

In [4]:
data.head()

age  sex  cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0  52.0    1   0     125.0  212.0  0.0        1    168.0    0.0      1.0   
1  53.0    1   0       NaN  203.0  1.0        0    155.0    1.0      3.1   
2  70.0    1   0     145.0    NaN  0.0        1    125.0    1.0      2.6   
3  61.0    1   0     148.0  203.0  0.0        1    161.0    0.0      0.0   
4  62.0    0   0     138.0  294.0  1.0        1      NaN    0.0      1.9   

   slope  ca  thal  target  
0      2   2   3.0     0.0  
1      0   0   3.0     0.0  
2      0   0   3.0     0.0  
3      2   1   3.0     0.0  
4      1   3   2.0     0.0

Кодирование категориальных признаков

In [5]:
data['slope'] = data['slope'].astype('category')
data['ca'] = data['ca'].astype('category')
data['thal'] = data['thal'].astype('category')

Назначение закодированной переменной новому столбцу

In [6]:
data['slope_category'] = data['slope'].cat.codes
data['ca_category'] = data['ca'].cat.codes
data['thal_category'] = data['thal'].cat.codes

In [7]:
data = data.drop(['slope', 'ca', 'thal'], axis=True)

In [8]:
data.head()

age  sex  cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0  52.0    1   0     125.0  212.0  0.0        1    168.0    0.0      1.0   
1  53.0    1   0       NaN  203.0  1.0        0    155.0    1.0      3.1   
2  70.0    1   0     145.0    NaN  0.0        1    125.0    1.0      2.6   
3  61.0    1   0     148.0  203.0  0.0        1    161.0    0.0      0.0   
4  62.0    0   0     138.0  294.0  1.0        1      NaN    0.0      1.9   

   target  slope_category  ca_category  thal_category  
0     0.0               2            2              3  
1     0.0               0            0              3  
2     0.0               0            0              3  
3     0.0               2            1              3  
4     0.0               1            3              2

In [9]:
data.isna().sum()

age                3
sex                0
cp                 0
trestbps          17
chol              18
fbs                4
restecg            0
thalach           10
exang              2
oldpeak            0
target             1
slope_category     0
ca_category        0
thal_category      0
dtype: int64

In [10]:
data=data.dropna()
data.isna().sum()

age               0
sex               0
cp                0
trestbps          0
chol              0
fbs               0
restecg           0
thalach           0
exang             0
oldpeak           0
target            0
slope_category    0
ca_category       0
thal_category     0
dtype: int64

In [11]:
data.describe().T

count        mean        std    min    25%    50%    75%  \
age             971.0   54.349125   9.089965   29.0   47.0   55.0   61.0   
sex             971.0    0.699279   0.458808    0.0    0.0    1.0    1.0   
cp              971.0    0.937178   1.028044    0.0    0.0    1.0    2.0   
trestbps        971.0  131.578785  17.636850   94.0  120.0  130.0  140.0   
chol            971.0  245.492276  51.554462  126.0  211.0  239.0  275.0   
fbs             971.0    0.148301   0.355581    0.0    0.0    0.0    0.0   
restecg         971.0    0.528321   0.527561    0.0    0.0    1.0    1.0   
thalach         971.0  149.063852  23.137466   71.0  132.0  152.0  166.0   
exang           971.0    0.336766   0.472848    0.0    0.0    0.0    1.0   
oldpeak         971.0    1.056643   1.170569    0.0    0.0    0.8    1.6   
target          971.0    0.518023   0.499933    0.0    0.0    1.0    1.0   
slope_category  971.0    1.391349   0.610296    0.0    1.0    1.0    2.0   
ca_category     971.0    0.761071   1.035737    0.0    0.0    0.0    1.0   
thal_category   971.0    2.317199   0.627795   -1.0    2.0    2.0    3.0   

                  max  
age              77.0  
sex               1.0  
cp                3.0  
trestbps        200.0  
chol            564.0  
fbs               1.0  
restecg           2.0  
thalach         202.0  
exang             1.0  
oldpeak           6.2  
target            1.0  
slope_category    2.0  
ca_category       4.0  
thal_category     3.0

In [12]:
y=data['sex']
x=data.drop('sex', axis=True)

In [13]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state=45)

In [14]:
x_train.shape, y_train.shape

((728, 13), (728,))

In [15]:
x_test.shape, y_test.shape

((243, 13), (243,))

In [16]:
np.unique(y_train)

array([0, 1], dtype=int64)

In [17]:
np.unique(y_test)

array([0, 1], dtype=int64)

Метрики построенной модели

In [18]:
scaler=MinMaxScaler().fit(x_train)
x_train=pd.DataFrame(scaler.transform(x_train), columns=x_train.columns)
x_test=pd.DataFrame(scaler.transform(x_test), columns=x_test.columns)
x_train.describe()

age          cp    trestbps        chol         fbs     restecg  \
count  728.000000  728.000000  728.000000  728.000000  728.000000  728.000000   
mean     0.529132    0.318223    0.357415    0.269058    0.157967    0.262363   
std      0.192017    0.346927    0.164923    0.119793    0.364961    0.261959   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.375000    0.000000    0.245283    0.188927    0.000000    0.000000   
50%      0.562500    0.333333    0.339623    0.253425    0.000000    0.500000   
75%      0.666667    0.666667    0.433962    0.335616    0.000000    0.500000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

          thalach       exang     oldpeak      target  slope_category  \
count  728.000000  728.000000  728.000000  728.000000      728.000000   
mean     0.593071    0.324176    0.175470    0.521978        0.690934   
std      0.176719    0.468388    0.194880    0.499860        0.309101   
min      0.000000    0.000000    0.000000    0.000000        0.000000   
25%      0.465649    0.000000    0.000000    0.000000        0.500000   
50%      0.618321    0.000000    0.129032    1.000000        0.500000   
75%      0.725191    1.000000    0.290323    1.000000        1.000000   
max      1.000000    1.000000    1.000000    1.000000        1.000000   

       ca_category  thal_category  
count   728.000000     728.000000  
mean      0.189560       0.830014  
std       0.262489       0.159097  
min       0.000000       0.000000  
25%       0.000000       0.750000  
50%       0.000000       0.750000  
75%       0.250000       1.000000  
max       1.000000       1.000000

In [19]:
def test_model(model):
    print("mean_absolute_error: ", mean_absolute_error(y_test, model.predict(x_test)))
    print("mean_squared_error: ", mean_squared_error(y_test, model.predict(x_test)))
    print("median_absolute_error: ", median_absolute_error(y_test, model.predict(x_test)))
    print("r2_score: ", r2_score(y_test, model.predict(x_test)))

Метод ближайших соседей с гиперпараметром K=2

In [20]:
reg_2=KNeighborsRegressor(n_neighbors=2)
reg_2.fit(x_train, y_train)

KNeighborsRegressor(n_neighbors=2)

In [21]:
test_model(reg_2)

mean_absolute_error:  0.06584362139917696
mean_squared_error:  0.043209876543209874
median_absolute_error:  0.0
r2_score:  0.7855882352941177


mean_absolute_error: 0.066, чем ближе к нулю, тем выше качество регрессии

mean_squared_error:  0.043, чем ближе к нулю, тем лучше модель

median_absolute_error:  0.0

r2_score:  0.786, чем ближе к 0, тем слабее зависимость, то есть разброс предсказаний модели  y относительно разброса самой целевой переменной y достаточно велик - то есть модель плохо описывает данные